File: 834-sum-of-distances-in-tree/834-sum-of-distances-in-tree.java
class Solution {
    int[] ans, count;
    List<List<Integer>> graph;
    int N;
    public int[] sumOfDistancesInTree(int n, int[][] edges) {
        this.N = n;
        ans = new int[N];
        count = new int[N];
        Arrays.fill(count, 1);
        graph = new ArrayList<>();
        for(int i = 0; i < n; i++) graph.add(new ArrayList<>());
        
        for(int[] edge : edges){
            int u = edge[0], v = edge[1];
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        
        dfs(0, new HashSet<Integer>());
        dfs2(0, new HashSet<Integer>());
        return ans;
    }
    
     public void dfs(int node, Set<Integer> visited) {
        visited.add(node);
        for (int child: graph.get(node))
            if(!visited.contains(child)){
                dfs(child, visited);
                count[node] += count[child];
                ans[node] += ans[child] + count[child];
            }
    }

    public void dfs2(int node, Set<Integer> visited) {
        visited.add(node);
        for (int child: graph.get(node))
            if (!visited.contains(child)) {
                ans[child] = ( ans[node] - count[child] ) + N - count[child];
                dfs2(child, visited);
            }
    }
}


/*
st: 1:17 

Keywords : undirected, n nodes, n - 1 edges 


answer[i] = sum of distances between ith node in the tree and all other nodes 



*/

File: 834-sum-of-distances-in-tree/834-sum-of-distances-in-tree.py
class Solution:
    def sumOfDistancesInTree(self, n: int, edges: List[List[int]]) -> List[int]:
        sub_tree = [0] * n 
        sum_dist = [0] * n 
        
        tree = defaultdict(list) 
        for u,v in edges:
            tree[u].append(v) 
            tree[v].append(u) 
        
        def dfs1(node, p):
            sub_tree[node] = 1 
            for adj in tree[node]:
                if adj != p:
                    dfs1(adj,node) 
                    sub_tree[node] += sub_tree[adj] 
                    sum_dist[node] += (sum_dist[adj] + sub_tree[adj])
        def dfs2(node,p):
            for adj in tree[node]:
                if adj != p:
                    adj_size = sub_tree[adj] 
                    out_side_adj = n - adj_size 
                    adj_contb = sum_dist[adj] + adj_size
                    sum_dist[adj] += (sum_dist[node] - adj_contb) + out_side_adj
                    dfs2(adj,node) 
                    
        dfs1(0,-1) 
        # print(sub_tree)
        # print(sum_dist)
        dfs2(0,-1)
        # print(sum_dist)
            
        return sum_dist